cpp/dolfinx/common/sort.h

```diff
@@ -23,144 +27,81 @@ namespace dolfinx
-  // maxvalue
-  int its = 0;
-  while (range)
-  {
-    range >>= BITS;
-    its++;
-  }
+    T max_value = proj(*std::ranges::max_element(range, std::less{}, proj));
 
-  // Adjacency list arrays for computing insertion position
-  std::array<std::int32_t, bucket_size> counter;
-  std::array<std::int32_t, bucket_size + 1> offset;
-
-  std::vector<std::int32_t> perm2(perm.size());
-  std::span<std::int32_t> current_perm = perm;
-  std::span<std::int32_t> next_perm = perm2;
-  for (int i = 0; i < its; i++)
-  {
-    // Zero counter
-    std::ranges::fill(counter, 0);
+    // Sort N bits at a time
+    constexpr int bucket_size = 1 << BITS;
+    T mask = (T(1) << BITS) - 1;
 
-    // Count number of elements per bucket
-    for (auto cp : current_perm)
+    // Compute number of iterations, most significant digit (N bits) of
+    // maxvalue
+    int its = 0;
+    while (max_value)
     {
-      T value = array[cp] - *min;
-      std::int32_t bucket = (value & mask) >> mask_offset;
-      counter[bucket]++;
+      max_value >>= BITS;
+      its++;
     }
 
-    // Prefix sum to get the inserting position
-    offset[0] = 0;
-    std::partial_sum(counter.begin(), counter.end(), std::next(offset.begin()));
+    // Adjacency list arrays for computing insertion position
+    std::array<std::int32_t, bucket_size> counter;
+    std::array<std::int32_t, bucket_size + 1> offset;
 
-    // Sort py permutation
-    for (auto cp : current_perm)
+    std::int32_t mask_offset = 0;
+    std::vector<T> buffer(range.size());
+    std::span<T> current_perm = range;
+    std::span<T> next_perm = buffer;
+    for (int i = 0; i < its; i++)
     {
-      T value = array[cp] - *min;
-      std::int32_t bucket = (value & mask) >> mask_offset;
-      std::int32_t pos = offset[bucket + 1] - counter[bucket];
-      next_perm[pos] = cp;
-      counter[bucket]--;
+      // Zero counter array
+      std::ranges::fill(counter, 0);
+
+      // Count number of elements per bucket
+      for (T c : current_perm)
+        counter[(proj(c) & mask) >> mask_offset]++;
+
+      // Prefix sum to get the inserting position
+      offset[0] = 0;
+      std::partial_sum(counter.begin(), counter.end(),
+                       std::next(offset.begin()));
+      for (T c : current_perm)
+      {
+        std::int32_t bucket = (proj(c) & mask) >> mask_offset;
+        std::int32_t new_pos = offset[bucket + 1] - counter[bucket];
+        next_perm[new_pos] = c;
+        counter[bucket]--;
+      }
+
+      mask = mask << BITS;
+      mask_offset += BITS;
+
+      std::swap(current_perm, next_perm);
     }
 
-    std::swap(current_perm, next_perm);
-
-    mask = mask << BITS;
-    mask_offset += BITS;
+    // Copy data back to array
+    if (its % 2 != 0)
+      std::ranges::copy(buffer, range.begin());
   }
+};
 
-  if (its % 2 == 1)
-    std::ranges::copy(perm2, perm.begin());
-}
+/// Radix sort
+inline constexpr __radix_sort radix_sort{};
```

Comment:
Is this `inline` necessary?

Reply:
I believe so - every time we include the `sort.h` we will have a different definition of `radix_sort` so we must allow for multiple definitions.